be/src/exec/spill/input_stream.h

```diff
@@ -74,28 +73,75 @@ class YieldableRestoreTask {
     std::vector<SpillInputStream*> _sub_stream;
 };
 
-// Note: not thread safe
+// Group of Blocks.
+// If spill needs sorted block output. Data within a block group is usually ordered.
+// Multiple block groups need to be merged and sorted.
+// Not thread safe.
+//
 class BlockGroup {
 public:
     BlockGroup() = default;
 
     void append(BlockPtr block) {
-        std::lock_guard guard(_mutex);
+        DCHECK(block != nullptr);
+        _data_size = std::nullopt;
         _blocks.emplace_back(std::move(block));
     }
 
+    const std::vector<BlockPtr>& blocks() { return _blocks; }
+
+    size_t data_size() const {
+        if (_data_size.has_value()) {
+            return _data_size.value();
+        }
+        size_t data_size = 0;
+        for (const auto& block : _blocks) {
+            data_size += block->size();
+        }
+        _data_size = data_size;
+        return _data_size.value();
+    }
+    size_t num_rows() const {
+        size_t num_rows = 0;
+        for (const auto& block : _blocks) {
+            num_rows += block->num_rows();
+        }
+        return num_rows;
+    }
+
+private:
+    mutable std::optional<size_t> _data_size;
```

Comment:
why use std::optional to store data size?

Reply:
the optional is useful to cache the data_size. A block group can have thousands of blocks.
```
    size_t data_size() const {
        if (_data_size.has_value()) {
            return _data_size.value();
        }
        size_t data_size = 0;
        for (const auto& block : _blocks) {
            data_size += block->size();
        }
        _data_size = data_size;
        return _data_size.value();
    }
```

Reply:
comments here